### PR TITLE
Avoid overwriting dependencies with different markers in `uv add`

### PR DESCRIPTION
**Pull request by blueraft**
## Summary

Splitting out https://github.com/astral-sh/uv/pull/6005

## Test Plan

`cargo test`

### CODE REVIEW

File: crates/uv/src/commands/project/remove.rs [hunk 1]
```diff
@@ -93,29 +95,30 @@ pub(crate) async fn remove(
     }?;
 
     for package in packages {
+        let requirement = Requirement::from_str(package.as_str())?;
```
Comment (blueraft) [inline crates/uv/src/commands/project/remove.rs:98]: I don't know if we can avoid doing this conversion here. All of the `remove_dependency` functions rely on the `find_dependencies` function which requires the `Requirement` type. `\cc @zanieb

Comment (zanieb) [reply crates/uv/src/commands/project/remove.rs:98]: From a quick scan of the changes, it seems like there are a couple alternative options:

1. Perform marker filtering on the result of `find_dependencies` when relevant or...
2. Update `find_dependencies` to take `Option<Marker>` in addition to `PackageName`

**Comment by zanieb** [comment]
Not sure if this is covered, but we'll probably want `uv remove foo` to remove all dependencies on `foo` regardless of the markers?

**Comment by ibraheemdev** [comment]
Haven't read the PR in detail yet, but if there is only a single instance of the package name can we update it regardless of the markers (without overwriting them if unspecified)? For example, `uv add requests; os_name == 'Linux'` followed by `uv add requests==2.0.0`: should the original dependency be updated or should we add a new entry?

**Comment by zanieb** [comment]
>  For example, uv add requests; os_name == 'Linux' followed by uv add requests==2.0.0: should the original dependency be updated or should we add a new entry?

Hard choice here — but if we update the existing entry here then there's no way to add a dependency _without_ a marker so I think we need to add a new entry.

File: crates/uv/src/commands/project/remove.rs (hunk 1 above)
Comment (blueraft) [reply crates/uv/src/commands/project/remove.rs]: I've gone with the second option

**Comment by blueraft** [comment]
> Not sure if this is covered, but we'll probably want uv remove foo to remove all dependencies on foo regardless of the markers?

It should be covered now. 

> For example, uv add requests; os_name == 'Linux' followed by uv add requests==2.0.0: should the original dependency be updated or should we add a new entry?

I agree with Zanie. If we were to update, we would also have to consider version comparisons before updating. `uv add requests==2.30; os_name == 'Linux'` followed by `uv add requests>=2.0.0` should create an additional entry.

File: crates/uv/src/commands/project/remove.rs [hunk 2]
```diff
@@ -223,18 +223,20 @@ enum Target {
 ///
 /// This is useful when a dependency of the user-specified type was not found, but it may be present
 /// elsewhere.
-fn warn_if_present(name: &PackageName, pyproject: &PyProjectTomlMut) {
-    for dep_ty in pyproject.find_dependency(name) {
+fn warn_if_present(requirement: &PackageName, pyproject: &PyProjectTomlMut) {
```
Comment (zanieb) [inline crates/uv/src/commands/project/remove.rs:226]: Can you revert the variable changes here where it's a `PackageName` again?

Comment (blueraft) [reply crates/uv/src/commands/project/remove.rs]: Done

**Comment by blueraft** [comment]
Hm something went wrong with Github Actions there, why is it building the docker image? 🤔

**Review by zanieb — APPROVED** [review]
Thanks!

**Pull request merged**